### ISSUE TITLE
SSLV3_ALERT_HANDSHAKE_FAILURE

### ISSUE THREAD

**Issue by kev946** [wildfoundry/dataplicity-lomond#32]
I am trying to connect to `wss://ws-feed-public.sandbox.gdax.com`, but get this error:

```bash
Connecting(url='wss://ws-feed-public.sandbox.gdax.com')
ConnectFail('[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:661)')
```

This is the version of python ssl:
```python
>>> import ssl
>>> ssl.OPENSSL_VERSION
'OpenSSL 1.0.2g  1 Mar 2016'
```

Here is the code I am using:

```python
from lomond import WebSocket

websocket = WebSocket('wss://ws-feed-public.sandbox.gdax.com')
for event in websocket:
    print(event)
```

Do you know could be happening here?

**Comment by willmcgugan** [comment]
That error message comes from `socket.connect`.  Can you try the following?

```
pip install certifi
```

**Comment by willmcgugan** [comment]
Managed to reproduce that error here (certifi didn't help). I tried `wss://ws-feed.gdax.com` and that worked as expected, so I'd guess that the URL you are using isn't serving websockets.

**Comment by kev946** [comment]
The  [websocket-client](https://github.com/websocket-client/websocket-client) library seems to be working:
```python
import websocket
def on_open(ws):
    print('open')
ws = websocket.WebSocketApp("wss://ws-feed-public.sandbox.gdax.com", on_open=on_open)
ws.run_forever()
```
this prints:

```bash
--- request header ---
GET / HTTP/1.1
Upgrade: websocket
Connection: Upgrade
Host: ws-feed-public.sandbox.gdax.com
Origin: http://ws-feed-public.sandbox.gdax.com
Sec-WebSocket-Key: w+CRuQQKh/+sjvnDWc6IhQ==
Sec-WebSocket-Version: 13


-----------------------
--- response header ---
HTTP/1.1 101 Switching Protocols
Date: Tue, 09 Jan 2018 15:34:03 GMT
Connection: upgrade
Set-Cookie: __cfduid=df3a8cc6c093f85d613727a670b2379751515512043; expires=Wed, 09-Jan-19 15:34:03 GMT; path=/; domain=.gdax.com; HttpOnly
Upgrade: websocket
Sec-WebSocket-Accept: bOyNl6icBXMbKks4lyr24qfi/Ag=
Sec-WebSocket-Version: 13
WebSocket-Server: uWebSockets
Strict-Transport-Security: max-age=15552000; includeSubDomains; preload
X-Content-Type-Options: nosniff
Server: cloudflare-nginx
CF-RAY: 3da8755facbe5619-ORD
-----------------------
open
```

**Comment by kev946** [comment]
I think I figured out what is going on. The server at `ws-feed-public.sandbox.gdax.com` implements [SNI](https://en.wikipedia.org/wiki/Server_Name_Indication) which requires a hostname to be sent at the start of the handshaking process. For example, we can connect like this:

```bash
openssl -connect ws-feed-public.sandbox.gdax.com -servername ws-feed-public.sandbox.gdax.com`
```
[ssl.wrap_socket](https://docs.python.org/3/library/ssl.html#ssl.wrap_socket) does not accept a `server_hostname` argument, but [ssl.SSLContext.wrap_socket](https://docs.python.org/3/library/ssl.html#ssl.SSLContext.wrap_socket) does. I can successfully connect by changing [the code here](https://github.com/wildfoundry/dataplicity-lomond/blob/master/lomond/session.py#L109) to:

```python
sock = ssl.SSLContext().wrap_socket(sock, server_hostname=self._address[0])
```

But this seems pretty hacky.

**Comment by willmcgugan** [comment]
Thanks for investigating. I did see there was some SNI code in websocket-client which is probably related to this. Will look in to a fix for that.

**Comment by willmcgugan** [comment]
Hi, We've implemented SNI support in Lomond. I got that service working with the following code:

```python
from lomond import WebSocket

subscribe_json = u"""
{
    "type": "subscribe",
    "product_ids": [
        "ETH-USD",
        "ETH-EUR"
    ],
    "channels": [
        "level2",
        "heartbeat",
        {
            "name": "ticker",
            "product_ids": [
                "ETH-BTC",
                "ETH-USD"
            ]
        }
    ]
}
"""

websocket = WebSocket('wss://ws-feed-public.sandbox.gdax.com')
for event in websocket:
    if event.name == "ready":
        websocket.send_text(subscribe_json)
    elif event.name == "text":
        print(event.text)
```

This works with lomond 0.1.13a0. I'll do a full release soon.

**Comment by kev946** [comment]
Sweet, thanks!

**Comment by willmcgugan** [comment]
Fixed in v0.1.13
